string_indexing/suffix_array.py

```diff
@@ -63,6 +66,223 @@ def induced_sorting(text, n):
   '''Computes suffix array using Nong-Zhang-Chan algorithm'''
   raise NotImplementedError
 
+def _map_string(s):
+  sorted_char_set = sorted(set( c for c in s[1:] ))
+  char_to_ord = { c : i for i, c in enumerate(sorted_char_set, 1) }
+  return [0] + [ char_to_ord[c] for c in s[1:] ] + [0]
+
+def small_large(text, n):
+  Category = enum.Enum("Category", "L S")
+
+  def _small_large(s):
+    A, SL, N = _bucket_array(s), _categorize_sl(s), len(s)-1
+    if N == 2:
+      return [2, 1]
+    if SL.count(Category.L) == 1:
+      return [N] + list(range(1, N))
+    if SL.count(Category.S) == 1:
+      return list(reversed(range(1, N + 1)))
+
+    if SL[-1] == Category.S:
+      sortedS = _sort_category(s, SL, Category.S)
+      groups = A.groups.copy()
+      _move_to_group_end(
```

Comment:
Here I was wondering about using the named parameter convention to increase readability e.g. it's not clear what True/False means without looking at the function signature. But please treat it as completely optional nit.